[JENKINS-20272] - Disconnected nodes should not be disconnected repeatedly

olivergondza opened the pull request:
See [JENKINS-20272](https://issues.jenkins-ci.org/browse/JENKINS-20272).

I consider the original fix fix to have no effect as [described in the jira](https://issues.jenkins-ci.org/browse/JENKINS-20272?focusedCommentId=338428&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-338428). Contrary to the original approach of skipping the check (that was not performed even before), I am focusing on the processing part that, as I was able to demonstrate, turns the absence of data caused by skipped monitoring into a failure its accumulation causes the monitor to consider the computer irresponsive.

### Proposed changelog entries

* Do not repeatedly disconnect offline computers.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

Contributors of the original PR (https://github.com/jenkinsci/jenkins/pull/2911): @oleg-nenashev, @abayer, @stephenc, @daniel-beck

olivergondza committed:
[JENKINS-20272] Reproduce in test

olivergondza committed:
[FIX JENKINS-20272] Do not disconnect skipped computers

jeffret-b reviewed (commented):
I think this is a useful change, which improves certain situations. I really like how the test demonstrates the situation it is addressing. Certainly makes it much easier to understand and evaluate.

The approach looks good, though I have questions about a few details. I also notice there are a few typos in comments. I can point those out if that is helpful.

jeffret-b commented on core/src/main/java/hudson/node_monitors/AbstractAsyncNodeMonitorDescriptor.java:
```diff
@@ -101,11 +111,46 @@ protected T monitor(Computer c) throws IOException, InterruptedException {
     private static final Logger LOGGER = Logger.getLogger(AbstractAsyncNodeMonitorDescriptor.class.getName());
+
+    /**
+     * Crate returned by {@link AbstractAsyncNodeMonitorDescriptor#monitor()}, it extends Map as that is what the interface
```
It just struck me that this is probably another typo. I think "Crate" means "Create", though the rest of the line is still a little confusing.

olivergondza replied:
Heh, my bad. It seems the "Crate design pattern" has made its way to no more than a single publication on design patterns (not even a good one) without wider adoption.

The class will serve as a "bag" to get several conceptual values through an interface where single value is expected/desired allowing for future evolution of the API without breaking the method API.

I will rephrase that, suggestions welcome.

jeffret-b replied:
Devin and I wondered about something like that, but I couldn't find any reference to that sort of usage anywhere. I didn't look very hard so that must be why I missed that one instance. British English may be different from my American English, but I've never heard "crate" used in that sort of form.

An alternative word might be "container", as in "Container returned by ...". Or if you want to maintain the naming from that one design pattern, it might be clearer as "Crate (or container)". 

To be really picky, there should be different punctuation on that line. Something like, "Crate returned by {@link AbstractAsyncNodeMonitorDescriptor#monitor()}. It extends Map as that is what the interface ...". Or possibly, "Crate returned by {@link AbstractAsyncNodeMonitorDescriptor#monitor()}; it extends Map as that is what the interface ...". I'm not worried about this bit, though. The meaning is still plenty clear as it is, but since we were talking about this line, I thought I'd mention it.

olivergondza commented on core/src/main/java/hudson/node_monitors/ResponseTimeMonitor.java:
```diff
@@ -46,20 +46,23 @@
                 Computer c = e.getKey();
                 Data d = e.getValue();
+                if (base.getSkipped().contains(c)) {
+                    assert d == null;
```
> ... asserts are useful during development cycles but don't belong in production code.

That is the reason why Java they are turned on during unit testing and off in production.

My point here is to express expectation the computer in skipped set must have `null` as monitoring value. It will help when when eventual bug in the code decides to manifest its presence :P.

jeffret-b replied:
I'm fine with that. I just wanted to check if you thought it would help with anything.

olivergondza commented on core/src/main/java/hudson/node_monitors/AbstractAsyncNodeMonitorDescriptor.java:
```diff
@@ -101,11 +111,46 @@ protected T monitor(Computer c) throws IOException, InterruptedException {
+         * Computers that ware skipped during monitoring as they either do not have a a channel (offline) or the monitor
+         * have not produced the Callable. Computers that caused monitor to throw exception are not returned here.
+         */
+        protected @Nonnull List<Computer> getSkipped() {
```
I do not  see how such change would affect encapsulation or improve things in any way. The value returned from `monitor` call currently provides a map (Computer->T) with the data and not `T getResult(Computer)` so we are following the same pattern here. I find current approach more versatile as it permit clients to be implemented in generally more efficient way (thinking of `Collection#retainAll` and friends).

Put differently, I do not see much to gain to be this restrictive, I do not see much to loose to give users all the data we have so I lean towards the latter so we do not have to compromise on performance or extend the API next time someone needs to consume this.

olivergondza committed:
[JENKINS-20272] Clean the API a bit

olivergondza commented:
I have cleaned the API a bit:

- No more abusing of the Map contract to funnel out some more data.
- Hence no unchecked cast.
- Easier to understand Result object.
- Fixed javadoc.

oleg-nenashev reviewed (approved):
👍

jeffret-b reviewed (approved):
I like the cleanup. This looks good. Thanks!
